dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/startup/TableAlteror.java

```diff
@@ -1569,33 +1569,30 @@ private void upgradeImplicitAverageMonitoringRules()
 
         Map<Integer, List<Integer>> idMap = new HashMap<>();
 
-        try
-        {
-            Statement statement = holder.getStatement();
+        Statement statement = holder.getStatement();
 
-            for ( Integer distinctId : distinctIds )
-            {
-                List<Integer> foreignIds = new ArrayList<>();
+        for ( Integer distinctId : distinctIds )
+        {
+            List<Integer> foreignIds = new ArrayList<>();
 
-                ResultSet resultSet = statement.executeQuery( "SELECT " + col2 + " FROM " + table + " WHERE " + col1
-                    + "=" + distinctId );
+            String query = "SELECT " + col2 + " FROM " + table + " WHERE " + col1 + "=" + distinctId;
 
+            try (ResultSet resultSet = statement.executeQuery( query ))
+            {
                 while ( resultSet.next() )
                 {
                     foreignIds.add( resultSet.getInt( 1 ) );
                 }
 
                 idMap.put( distinctId, foreignIds );
             }
+            catch ( Exception ex )
+            {
+                log.error( ex );
+            }
         }
-        catch ( Exception ex )
-        {
-            log.error( ex );
-        }
-        finally
-        {
-            holder.close();
-        }
+
+        holder.close();
 

```

Comment:
I'm not sure if we should do this, perhaps revert, we need to ensure that "close()" is always called. My old StatementHolder stuff I think needs itself to implement Closeable as the holder.close() operation is closing the underlying streams.